contracts/serverless-v2/Executors.sol

```diff
@@ -0,0 +1,470 @@
+        uint256 removedAmount,
+        uint256 remainingStakedAmount
+    );
+
+    error ExecutorsLessStakeAmount();
+    error ExecutorsInvalidSigner();
+    error ExecutorsExecutorAlreadyExists();
+    error ExecutorsAlreadyDeregistered();
+    error ExecutorsInvalidAmount();
+    error ExecutorsInvalidExecutor();
+
+    //-------------------------------- internal functions start ----------------------------------//
+
+    function _registerExecutor(
+        bytes memory _attestationSignature,
+        IAttestationVerifier.Attestation memory _attestation,
+        uint256 _jobCapacity,
+        bytes memory _signature,
+        uint256 _stakeAmount
+    ) internal {
+        if(_stakeAmount < MIN_STAKE_AMOUNT)
+            revert ExecutorsLessStakeAmount();
+
+        // attestation verification
+        _verifyEnclaveKey(_attestationSignature, _attestation);
+
+        address operator = _msgSender();
+        // signature check
+        bytes32 hashStruct = keccak256(
+            abi.encode(
+                REGISTER_TYPEHASH,
+                operator,
+                _jobCapacity
+            )
+        );
+        bytes32 digest = keccak256(abi.encodePacked("\x19\x01", DOMAIN_SEPARATOR, hashStruct));
+        address signer = digest.recover(_signature);
+
+        _allowOnlyVerified(signer);
+
+        address enclaveKey = _pubKeyToAddress(_attestation.enclavePubKey);
+        if(signer != enclaveKey)
+            revert ExecutorsInvalidSigner();
+
+        // transfer stake
+        TOKEN.safeTransferFrom(operator, address(this), _stakeAmount);
+
+        if(executors[operator].enclaveKey != address(0))
+            revert ExecutorsExecutorAlreadyExists();
+        executors[operator] = Executor({
+            enclaveKey: enclaveKey,
+            jobCapacity: _jobCapacity,
+            activeJobs: 0,
+            stakeAmount: _stakeAmount,
+            status: true,
+            unstakeStatus: false,
+            unstakeAmount: 0
+        });
+
+        // add node to the tree
+        _insert_unchecked(operator, uint64(_stakeAmount));
+
+        emit ExecutorRegistered(operator, enclaveKey);
+    }
+
+    function _deregisterExecutor() internal {
+        address operator = _msgSender();
+        _isValidExecutor(operator);
+        if(!executors[operator].status)
+            revert ExecutorsAlreadyDeregistered();
+
+        executors[operator].status = false;
+
+        if(executors[operator].activeJobs == 0) {
+            // return stake amount
+            TOKEN.safeTransfer(operator, executors[operator].stakeAmount);
+            _revokeEnclaveKey(executors[operator].enclaveKey);
+            delete executors[operator];
+        }
+
+        // remove node from the tree
+        _deleteIfPresent(operator);
+
+        emit ExecutorDeregistered(operator);
+    }
+
+    function _addExecutorStake(
+        uint256 _amount
+    ) internal {
+        if(_amount == 0)
+            revert ExecutorsInvalidAmount();
```

Comment:
do we care

Reply:
just to prevent 0 amount txns, although no profit for the executor. Can remove if not so important